### ISSUE TITLE
[clang] an issue with compound literals ?

### ISSUE THREAD

**Issue by swordkill** [llvm/llvm-project#107562]
**Hi Humans,**

> Last year I shared an issue related to VLA and Blocks (extension) :
https://github.com/llvm/llvm-project/issues/63412

I come back for one more embarrassing, related either to C99 compound literals, or C11 _Generics or _Static_assert.

Here is a reproduction : 
https://godbolt.org/z/9dTa4YP4n

### The purpose is to be able to use static assertion within an expression.

```c
/*
** Using -std=c23.
**
** The trick is that static_assert is garanteed by the standard 
** to not produce anything :
** 
** - https://en.cppreference.com/w/c/language/_Static_assert
**
** """Otherwise, 
**      if expression does not equal zero, nothing happens; 
**      no code is emitted.
** """
**/
#define static_expr(e) \
    ( !! sizeof (struct {static_assert (e); char c;}) )

#define is_same(e1, e2)                 \
    _Generic (&(typeof_unqual (e1)) {}  \
        , typeof_unqual (e2) * : true   \
        , default              : false  \
    )
```

### The issue

```c
/*
** It's OK.
**/
typeof        ((size_t) {i}) witness_one;
typeof_unqual ((size_t) {i}) witness_two;

/*
** It's OK.
**/
is_same (witness_one, (size_t) {i});
is_same (witness_two, (size_t) {i});

/*
** It's OK.
**/
static_assert (is_same (witness_one, (size_t) {i}));

/*
** It's KO.
** [clang] error: initializer element is not a compile-time constant
** [gcc  ] everything is good buddy
**/
static_expr (is_same (witness_one, (size_t) {i}));
```

**Comment by swordkill** [comment]
For short :

```c
// Changing 'int i' to 'const int i' makes it work.

// 1 : sizeof does not care knowing the value, it needs only the type.
// 2 : static_assert is guaranteed by the standard to not generate anything when expression is true.

// By 1 and 2 : (supposing sizeof size_t is 8 bytes)
//     static_assert (!! sizeof ((size_t) {i}));
// --> static_assert (!! 8)
// --> static_assert (true)
// and struct { static_assert (true) }
// --> struct { }
// but [clang] error: initializer element is not a compile-time constant
// and [ gcc ] everything is ok buddy

/* const */ size_t i;

         static_assert (!! sizeof ((size_t) {i}));    // OK
struct { static_assert (!! sizeof ((size_t) {i})); }; // KO
```

**Comment by shafik** [comment]
CC @AaronBallman

**Comment by AaronBallman** [comment]
C23 6.5.3.6p3: All the constraints for initializer lists in 6.7.11 also apply to compound literals.
C23 6.7.11p5: All the expressions in an initializer for an object that has static or thread storage duration or is declared with the constexpr storage-class specifier shall be constant expressions or string literals.

So we're fine there (nothing else in 6.7.11's constraints really applies). I don't see any further constraints in 6.5.3.6 that would justify Clang's behavior, so confirming the issue.

**Comment by zygoloid** [comment]
I don't have C23 to hand, but in the draft I do have, I found this:

6.5.3.6/3: If the compound literal occurs outside the body of a function, the object has static storage duration; otherwise, it has automatic storage duration associated with the enclosing block.

but conversely:

6.2.4/6: If an initialization is specified for the object, it is performed each time the declaration or compound literal is reached in the execution of the block; otherwise, the value becomes indeterminate each time the declaration is reached.

... which together seem to mean that the compound literal isn't required to have a constant initializer in this case, but also that its initializer is never actually evaluated at runtime either. Which is weird, but OK I guess, given that presumably it can only occur in things like `sizeof` or `typeof` anyway.

That behavior is also inconsistent:

```c
int n;

void f() {
  struct A {
    // OK.
    int array[sizeof((int){n})];
  };
}

struct B {
  // Ill-formed.
  int array[sizeof((int){n})];
};
```

... but accepting `A` and rejecting `B` is what GCC does, and what the C standard seems to ask for. *shrug*

**Comment by SwordKill** [comment]
The original code for my fun little project is written in C23, but I should have removed the `-std=c23` flag since the issue isn't related to that.

Compound literals are part of C99, and I wanted to combined them with C11's `_Static_assert`.
My goal was to validate macro parameters. This is achievable using a GNU extension  called `statement expression`, however I wanted to adhere to the standard.

To work around this, I noticed that C++ explicitly states that a static assertion can be a class member. Additionally, the C standard specifies that a static assertion that evaluates to true doesn't generate any code. So, I wrapped the static assertion in a structure with an additional char to avoid creating an empty struct, which isn't standard either. Finally, enclosing the entire structure in a sizeof would either trigger a compile-time error or evaluate to 1 (true).

This way, it can be used virtually anywhere.

To provide a concrete example, I have a generic function-like macro `vector_push(self, object)` that copies the `object` to the end of the `self` vector. This macro triggers a compile-time error if the `object` type does not match the type for which the `self` vector was declared.

For testing purposes, I then tried to push compound literals in a loop, like this:

```c
typedef struct
{
    int x;
    int y;
}
    Point;

typedef Vector (Point) VectorOfPoint;
```

Below, `vector` is a `VectorOfPoint`:

```c
for (int i = 4096; i --> 0;)
    vector_push (vector, (Point) {.x = i, .y = i}); // [clang] error: initializer element is not a compile-time constant
```

It works wonderfully with GCC. If I try to push anything other than a `Point` into the `VectorOfPoint`, the static assertion fails. Additionally, my editor with the `clangd` language server detects the issue in real-time.

In conclusion, I created a single `vector_push` method-like macro that works with any type of vector while still enforcing type safety. It's similar to a C++ template but implemented using standard C.

**Comment by SwordKill** [comment]
How dare you change my concise title? You're like those recent anime that spoil the entire story in the title! Shame on you! Ya I'm french and I've recently discovered ya don't put a space before stacked punctuation marks in english (like semi-colon, colon, interrogation and exclamation marks). That's the reason for ```SwordKill changed the title [clang] An issue with compound literals ? [clang] An issue with compound literals? 19 hours ago```.

**Comment by SwordKill** [comment]
Also, did you guys know that the C23 specification for the `auto` keyword is ambiguous? I mean, clearly it already behaves differently between Clang and GCC, with the GCC version being more restricted.

The paper (written by a French like me, so I’m not surprised it’s flawed) literally contradicts itself: [Type inference for object definitions](https://open-std.org/JTC1/SC22/WG14/www/docs/n3007.htm)

---
### Part 3 says:

`We propose standardizing the GNU C feature exactly, except for possibly changing the use of the extended specifier __auto_type to the existing specifier auto. The GNU C feature (since version 11) has clear semantics which can be expressed entirely in terms of the adopted typeof feature`.

### Then, Part 6.8 says:

`Such a declaration that also defines a structure or union type violates a constraint. Here, the identifier a which is not ordinary but in the name space of the structure type is declared.`

```c
AUTOTYPE p = (struct { int a; } *)0;
```
---

So, when mixing the comma operator with type inference, GCC produces:

```c
int  i = ( sizeof (struct {char c;}), 2 ); // ok
auto a = ( sizeof (struct {char c;}), 2 ); // [gcc] error: 'struct <anonymous>' defined in underspecified object initializer
```

So, GCC applies the restriction mentioned in Point 6.8, even though it contradicts Point 3, because __auto_type does not have such a constraint.

What about Clang? Well, Clang does not apply the restriction in 6.8 and correctly infers the type of the comma expression as that of its last operand.

I submitted a ticket to GCC Bugzilla about this issue: [Bugzilla](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=116631).

I ended up closing it because it couldn’t be classified as a bug. It’s more a matter of interpretation and arbitrary implementation decisions.

Take a look: [Compiler Explorer](https://godbolt.org/z/3xeovn9h4)

**Comment by SwordKill** [comment]
Well, this is off-topic, but it ultimately justifies the recent trend of adopting new languages with a single implementation (and therefore a single compiler). The divergence in C++20 and C23 implementations is so significant that they can hardly be called standards anymore, which also explains their poor adoption.

Just look at how C++20 modules are implemented so differently in MSVC, Clang, and GCC, and how C23 type inference behaves inconsistently across compilers.

It seems that MSVC also doesn’t apply the restriction in Point 6.8: [Compiler Explorer](https://godbolt.org/z/3fK3rbbb9)

In the end, what could have been a great evolution of the C language—specifically, the addition of powerful generic capabilities without the drawbacks of C++, such as a bloated specification and, more importantly, an unstable ABI—will likely fade away.

**Comment by AaronBallman** [comment]
> How dare you change my concise title? You're like those recent anime that spoil the entire story in the title! Shame on you!

Please familiarize yourself with our [Code of Conduct](https://llvm.org/docs/CodeOfConduct.html). For what it's worth, we frequently reword issue titles (and summaries) so that it's easier for community members to understand the issue without having to read through as many comments.

> Also, did you guys know that the C23 specification for the auto keyword is ambiguous? I mean, clearly it already behaves differently between Clang and GCC, with the GCC version being more restricted.

This is unrelated to the original issue; you should file a separate issue if you think Clang has a bug, otherwise it's too easy for things to get lost in discussion.

> Well, this is off-topic,

It is. :-)
